test/add/git/git_test.dart

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:pub/src/exit_codes.dart' as exit_codes;
+import 'package:test/test.dart';
+
+import '../../descriptor.dart' as d;
+import '../../test_pub.dart';
+
+void main() {
+  test('adds a package from git', () async {
+    ensureGit();
+
+    await d.git(
+        'foo.git', [d.libDir('foo'), d.libPubspec('foo', '1.0.0')]).create();
+
+    await d.appDir({}).create();
+
+    await pubAdd(args: ['foo', '--git-url', '../foo.git']);
```

Comment:
Can we add a test case that `pub add foo:^1.2.3 --git-url ../foo.git` won't work..

At-least I expect that one cannot have a version constraint if fetching from `git`, and that we have some special handling in place to ensure that the user gets an error message from `pub add`.

I haven't tested, but I assume that `dependencies: {'mypkg': {'git': {'url': '../foo.get', 'version': '1.2.3'}}}`  will never work in a `pubspec.yaml`, or at-least that the `version` constraint is ignored (as `pub` generally ignores unrecognized keys). I haven't tests, I'm just guessing that this is the case, we probably should test this too :D

Reply:
Nvm, version constraints are support even when source only has one version...